### PR TITLE
fix(finetune): shift logits / targets in val loss calc

### PR DESCRIPTION
**Pull request by MF-FOOM**
fixes https://github.com/Lightning-AI/lit-gpt/issues/407#issuecomment-1678507201

### CODE REVIEW

File: finetune/adapter.py [hunk 1]
```diff
@@ -218,7 +218,8 @@ def validate(
     for k in range(eval_iters):
         input_ids, targets = get_batch(fabric, val_data, longest_seq_length)
         logits = model(input_ids)
-        loss = chunked_cross_entropy(logits, targets, chunk_size=0)
+        logits[-1] = logits[-1][..., :-1, :]
+        loss = chunked_cross_entropy(logits, targets[..., 1:], chunk_size=0)
```
Comment (MF-FOOM) [inline finetune/adapter.py:222]: Should we be using `chunk_size=0` in the training loss calculation as well? It's currently unset there

Comment (MF-FOOM) [reply finetune/adapter.py:222]: Ah ok sounds like we shouldn't — chunk_size=0 is an optimization for the backward pass? (https://github.com/Lightning-AI/lit-gpt/pull/178/files#r1236187459)

Comment (MF-FOOM) [reply finetune/adapter.py:222]: Given this, shouldn't this `chunk_size=0` in `finetune.py`'s training loss calculation be removed? https://github.com/Lightning-AI/lit-gpt/blob/75bca51f7e86759f52ca258a9aee40922177de7f/finetune/full.py#L168

Comment (carmocca) [reply finetune/adapter.py:222]: chunk_size>0 only helps with memory peaks for models with frozen parameters, the one you linked doesn't freeze at all (hence the "full" name)

File: finetune/adapter.py [hunk 2]
```diff
@@ -218,7 +218,8 @@ def validate(
     for k in range(eval_iters):
         input_ids, targets = get_batch(fabric, val_data, longest_seq_length)
         logits = model(input_ids)
-        loss = chunked_cross_entropy(logits, targets, chunk_size=0)
+        logits[-1] = logits[-1][..., :-1, :]
```
Comment (carmocca) [inline finetune/adapter.py:221]: You don't need this list modification because `model.forward(lm_head_chunk_size=0)` by default. Instead, copy what's done for `full.py`

All the chunk size tricks are only relevant for `backward`, so you don't need them at validation

Comment (MF-FOOM) [reply finetune/adapter.py:221]: sorry not sure i follow?

Comment (MF-FOOM) [reply finetune/adapter.py:221]: you mean just use `logits[..., :-1, :], targets[..., 1:]`?

Comment (carmocca) [reply finetune/adapter.py:221]: Logits is only a list for the finetuning models when `lm_head_chunk_size>0` (not the default)

https://github.com/Lightning-AI/lit-gpt/blob/main/lit_gpt/lora.py#L532-L535

Which means that you want to do

```python
        chunked_cross_entropy(logits[..., :-1, :], targets[..., 1:], chunk_size=0)
```

as with `full.py`

**Comment by MF-FOOM** [comment]
should be good now, sorry about all this commits — doing this from the web browser atm lol

**Review by carmocca — APPROVED** [review]
LGTM! cc @awaelchli

**Review by awaelchli — APPROVED** [review]
What about the training loop, why are the logits for only the last chunk being shifted?
https://github.com/Lightning-AI/lit-gpt/blob/246c75109365d72aadc2a726553447efce4b6fef/finetune/lora.py#L191-L192

**Comment by carmocca** [comment]
> Why are the logits for only the last chunk being shifted?

Can you show what you'd expect? This is passing

```python
from lit_gpt.adapter import GPT

model = GPT.from_name("pythia-70m")
x = torch.randint(0, 8, (2, 64))

logits = model(x, lm_head_chunk_size=0)
logits = logits[..., :-1, :]

chunked_logits = model(x, lm_head_chunk_size=2)
assert len(chunked_logits) == 32
chunked_logits[-1] = chunked_logits[-1][..., :-1, :]
chunked_logits = torch.cat(chunked_logits, dim=1)

torch.testing.assert_close(logits, chunked_logits)
```

(merging since changes here would be unrelated to the current changes in this PR)

**Comment by awaelchli** [comment]
It's correct, truncating the last chunk is equivalent to shifting first and then chunking. Sorry for the false alarm.
